mitosheet/mitosheet/ai/prompt.py

```diff
@@ -0,0 +1,150 @@
+
+    if len(df_string_with_only_headers) < max_characters:
+        return df_string_with_only_headers
+
+    # Then, we find the maximum number of column headers we can include. Note that we use a binary search algorithm
+    # to do this - as the loop that we original started with had performance problems on dataframes with a large
+    # number of column headers that we noticed while testing
+    reduced_headers = ''
+    columns = df.columns.to_list()
+    low = 0
+    num_not_included = 1
+    high = len(columns) - 1
+    len_container = len(f'pd.DataFrame(columns=[])')
+
+    while low < high:
+        mid = (low + high) // 2
+        reduced_headers = column_header_list_to_transpiled_code(columns[:mid])
+        num_not_included = len(columns) - mid + 1
+
+        if (len(reduced_headers) + len_container) < max_characters:
+            low = mid + 1
+        elif (len(reduced_headers) + len_container) > max_characters:
+            high = mid - 1
+
+    return f'pd.DataFrame({OPEN_BRACKET}columns=[{reduced_headers} ... and {num_not_included} more]{CLOSE_BRACKET})'
+
+
+def get_input_data_string(df_names: List[str], dfs: List[pd.DataFrame], current_selection: Optional[Selection]) -> str:
+    # First, we get the input data code for creating the dataframes
+    dfs_creation_code = {}
+    for df_name, df in zip(df_names, dfs):
+        max_characters = MAX_CHARS_FOR_INPUT_DATA / len(df_names)
+        dfs_creation_code[df_name] = get_dataframe_creation_code(df.head(5), max_characters)
+
+    # Then, build the input data string, making sure to put the selected dataframe first - if these is one
+    if current_selection is not None:
+        input_data_string = f"""Input dataframes:
+```
+{current_selection['selected_df_name']} = {dfs_creation_code[current_selection['selected_df_name']]}
+{NEWLINE.join([f'{df_name} = {df_creation_code}' for df_name, df_creation_code in dfs_creation_code.items() if df_name != current_selection['selected_df_name']])}
+```
+The dataframe {current_selection['selected_df_name']} is currently selected.
+"""
+    else:
+        input_data_string = f"""Input dataframes:
+```
+{NEWLINE.join([f'{df_name} = {df_creation_code}' for df_name, df_creation_code in dfs_creation_code.items()])}
+```"""
+
+    return input_data_string
+
+def get_example_string(current_selection: Optional[Selection]) -> str:
+
+    # TODO: If there is a previous edit in the mitosheet, we could use this as the example, as it's about as in-context as you can get
+
+    # Otherwise, we make up a fake edit
+    if current_selection is not None:
+        return f"""Transformation: replace 1 with 2
+Code:
+```
+{current_selection['selected_df_name']}.replace(1, 2, inplace=True)
+```"""
+    else:
+        return f"""Transformation: replace 1 with 2 in df
+Code:
+```
+df.replace(1, 2, inplace=True)
+```"""
+
+
+def get_prompt(df_names: List[str], dfs: List[pd.DataFrame], current_selection: Optional[Selection], user_input: str) -> str:
```

Comment:
This only supports a single selection, but we let users have many selections. Do we want to support that? ie: "Apply this formula to all selected columns" could be a common command?

Reply:
The selection object on the backend is very different than frontend. We're just going to send all the selected columns and rows to the backend, and that will be considered one selection.